src/common/inbound.c

```diff
@@ -1901,6 +1932,90 @@ inbound_cap_list (server *serv, char *nick, char *extensions,
 								  NULL, NULL, 0, tags_data->timestamp);
 }
 
+static void plain_authenticate(server *serv, char *user, char *password)
+{
+	char *pass = encode_sasl_pass_plain (user, password);
+
+	if (pass == NULL)
+	{
+		/* something went wrong abort */
+		tcp_sendf (serv, "AUTHENTICATE *\r\n");
+		return;
+	}
+
+	/* long SASL passwords must be split into 400-byte chunks
+	   https://ircv3.net/specs/extensions/sasl-3.1#the-authenticate-command */
+	size_t pass_len = strlen (pass);
+	if (pass_len <= 400)
+		tcp_sendf (serv, "AUTHENTICATE %s\r\n", pass);
+	else
+	{
+		size_t sent = 0;
+		while (sent < pass_len)
+		{
+			char *pass_chunk = g_strndup (pass + sent, 400);
+			tcp_sendf (serv, "AUTHENTICATE %s\r\n", pass_chunk);
+			sent += 400;
+			g_free (pass_chunk);
+		}
+	}
+	if (pass_len % 400 == 0)
+		tcp_sendf (serv, "AUTHENTICATE +\r\n");
+}
+
+#ifdef USE_OPENSSL
+/*
+ * Sends AUTHENTICATE messages to log in via SCRAM.
+ */
+static void scram_authenticate(server *serv, const char *data, const char *digest,
+							   const char *user, const char *password)
+{
+	char *encoded, *decoded, *output;
+	int ret;
+	size_t output_len;
+	gsize decoded_len;
+
+	if (serv->scram_session == NULL) {
+		serv->scram_session = scram_create_session(digest, user, password);
+
+		if (serv->scram_session == NULL) {
+			// TODO: localized output
+			// g_error("Could not create SCRAM session with digest %s", digest);
+			tcp_sendf(serv, "AUTHENTICATE *\r\n");
+			return;
+		}
+	}
+
+	decoded = g_base64_decode(data, &decoded_len);
+	ret = scram_process(serv->scram_session, decoded, &output, &output_len);
+	g_free(decoded);
+
+	if (ret == SCRAM_IN_PROGRESS) {
+		// Authentication is still in progress
+		encoded = g_base64_encode((guchar *) output, output_len);
+		tcp_sendf(serv, "AUTHENTICATE %s\r\n", encoded);
+		g_free(encoded);
+		g_free(output);
+	} else if (ret == SCRAM_SUCCESS) {
+		// Authentication succeeded
+		tcp_sendf(serv, "AUTHENTICATE +\r\n");
+		scram_free_session(serv->scram_session);
+		serv->scram_session = NULL;
+	} else if (ret == SCRAM_ERROR) {
+		// Authentication failed
+		tcp_sendf(serv, "AUTHENTICATE *\r\n");
+
+		if (serv->scram_session->error != NULL) {
+			// TODO: localized output
+			// g_warning("SASL SCRAM authentication failed: %s", serv->scram_session->error);
```

Comment:
You could leave this as a `g_info()`, won't be shown by default but helpful for debugging. Same goes for any other prints you left behind if you think they're useful.

Reply:
Changed to g_info()